### ISSUE TITLE
Use your own model

### ISSUE THREAD

**Issue by Nazim0001** [SBU-BMI/wsinfer#221]
Hi,
I would like to use my own model with WSInfer. But I don't know how I need to do, and what kind of information I need to add in JSON file?
Could you please help me?
Best
Nazim

**Comment by kaczmarj** [comment]
hi @Nazim0001 - you can use one of the JSON files in the model zoo as a starting point, and replace things specific to your model. i have pasted the JSON for the [pancancer lymphocyte classification model](https://huggingface.co/kaczmarj/pancancer-lymphocytes-inceptionv4.tcga/blob/main/config.json). 

please note that the `architecture` key is unused but because i included it in the json schema, you need to include it. the value isn't used, so fill that in with anything you want.

you can validate your JSON against our schema using the following command:

```
wsinfer-zoo validate-config config.json
```

once you have the JSON file, you will also need to have your model in torchscript format. when you have your model and the config JSON, run the model using the following command:

```
wsinfer run --wsi-dir slides/ --results-dir results/ --model-path path/to/torchscript.pt --config config.json
```

here is an example config JSON:

```json
{
  "spec_version": "1.0",
  "architecture": "inception_v4nobn",
  "num_classes": 2,
  "class_names": [
    "Other",
    "Lymphocytes"
  ],
  "patch_size_pixels": 100,
  "spacing_um_px": 0.5,
  "transform": [
    {
      "name": "Resize",
      "arguments": {
        "size": 299
      }
    },
    {
      "name": "ToTensor"
    },
    {
      "name": "Normalize",
      "arguments": {
        "mean": [
          0.5,
          0.5,
          0.5
        ],
        "std": [
          0.5,
          0.5,
          0.5
        ]
      }
    }
  ]
}
```

**Comment by Nazim0001** [comment]
Thank you Jakub for your reply.

I thought that it would be possible to use WSInfer for a new task like to
quantify a staining, right?

**Comment by kaczmarj** [comment]
Could you describe your use case in more detail?Wsinfer does patch classification.Best,JakubOn Mar 17, 2024, at 10:39, Nazim0001 ***@***.***> wrote:﻿
Thank you Jakub for your reply.



I thought that it would be possible to use WSInfer for a new task like to

quantify a staining, right?



On Sun, Mar 17, 2024, 13:49 Jakub Kaczmarzyk, ***@***.***>

wrote:

**Comment by Nazim0001** [comment]
I would like to use wsinfer to count the number of cells with staining of
pdl1 over the number of total cells.

What do you think? Is it possible ?

**Comment by kaczmarj** [comment]
if you have a patch classification model that can do this, then yes, you could use wsinfer to run that on slides. the wsinfer model zoo does not include any models that count cells, so you would have to provide your own model.

but counting cells is not usually a patch classification task. you might want to consider other methods.

**Comment by ajr82** [comment]
Hi Jakub @kaczmarj 
I have used Wsinfer (built-in models) on QuPath and it is great. We would like to train and use custom models. 

Would pytorch models trained using Yolov8 be compatible with Wsinfer? These models have a .pt format. They can also be converted into Torchscript models. https://docs.ultralytics.com/modes/export/#arguments

**Comment by kaczmarj** [comment]
hi @ajr82 - if the models do multi-class classification or multi-label classification, then yes, they would be compatible. the YOLOv8 classification models might work after being converted to torchscript. wsinfer expects that the forward pass outputs a tensor of logits per patch.

from my experience with yolo models, some might also return a bounding boxes or segmentations. in that case, the wsinfer code would have to be modified.

do you only need classification? or are you trying to perform other tasks too?

**Comment by ajr82** [comment]
Thank you for getting back! 
Yes, Yolo models either return segmentations or bounding boxes. We were hoping to use these models to annotate specific structures in renal biopsies (like glomeruli, tubules) and eventually do morphometric analyzes.

**Comment by em4rtini** [comment]
dear all
I would like to use wsi infer to run my pytorch script model
My model is an image2image translation model
It takes a 3 channel 2048x2048 image and it translates it to a 1 channel 2048x2048 image

should it be possible with wsiInfer?
how the config.json should be?

thanks a lot

**Comment by kaczmarj** [comment]
hi @em4rtini - this wouldn't be possible with wsinfer, because wsinfer only performs patch-based classification. the inputs to wsinfer models are patches and outputs are classifications.
